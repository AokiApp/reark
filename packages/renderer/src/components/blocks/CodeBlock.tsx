import { BlockInnerComponent } from "../../types";
import { CODE_LANGUAGE } from "../../constants/codeLanguage";
import { Text } from "./Text";

export const CodeBlock: BlockInnerComponent = ({ block }) => {
  const code = block.code;
  if (!code) {
    return null;
  }
  const codeLanguage = code?.style?.language
    ? CODE_LANGUAGE[code.style.language]
    : "Unknown";

  return (
    <div className="reark-codeblock">
      <div className="reark-codeblock__language">{codeLanguage}</div>
      <pre className="reark-codeblock__pre">
        <Text elements={code.elements} style={code.style} />
      </pre>
    </div>
  );
};
// todo : copy button
