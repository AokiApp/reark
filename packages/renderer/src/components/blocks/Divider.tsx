import { BlockInnerComponent } from "../../types";

export const Divider: BlockInnerComponent = () => {
  return <hr className="reark-divider" />;
};
