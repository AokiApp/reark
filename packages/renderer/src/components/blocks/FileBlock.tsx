import { BlockInnerComponent } from "../../types";

/**
 * FileBlock: Renders a file block (block_type: 23).
 * - Shows file name and download link if available.
 * - This block is usually paired with a ViewBlock (block_type: 33) for visual representation.
 * - If rendered directly, provides basic file info.
 */
export const FileBlock: BlockInnerComponent = ({ block }) => {
  const file = block.file;
  if (!file) {
    return (
      <div className="reark-file-block reark-file-block--empty">
        No file data
      </div>
    );
  }

  return (
    <div className="reark-file-block">
      <div>
        <strong>File:</strong> {file.name || "(no name)"}
      </div>
      {file.token ? (
        <a
          href={`https://open.feishu.cn/drive/file/${file.token}`}
          target="_blank"
          rel="noopener noreferrer"
          className="reark-file-block__file-link"
        >
          Download / Open
        </a>
      ) : (
        <span className="reark-file-block__file-link--disabled">
          (No file token)
        </span>
      )}
    </div>
  );
};
