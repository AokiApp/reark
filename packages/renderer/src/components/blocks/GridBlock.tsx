import { BlockInnerComponent } from "../../types";
import { BlockComponent } from "../BlockComponent";

/**
 * GridBlock: Renders a grid container with a specified number of columns.
 * Each child is expected to be a GridColumnBlock.
 */
export const GridBlock: BlockInnerComponent = ({ block }) => {
  const grid = block.grid;
  if (!grid || !block.children || !Array.isArray(block.children)) {
    return null;
  }

  // column_size: number of columns (2-4)
  //const columnSize = grid.column_size || 2;

  // Render children by block reference (blockId), allowing for block reference resolution.
  return (
    <div className="reark-grid-wrapper">
      <div
        className="reark-grid"
        style={{
          display: "flex",
          //gridTemplateColumns: `repeat(${columnSize}, 1fr)`,
          gap: "16px",
        }}
      >
        {block.children.map((childId) => (
          // BlockComponent resolves and renders the block by its ID (block reference)
          <BlockComponent key={childId} blockId={childId} />
        ))}
      </div>
    </div>
  );
};
