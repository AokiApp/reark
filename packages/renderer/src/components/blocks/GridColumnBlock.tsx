import { BlockInnerComponent } from "../../types";
import { BlockComponent } from "../BlockComponent";

/**
 * GridColumnBlock: Renders a single column in a grid, with optional width ratio.
 * Children can be any block type.
 */
export const GridColumnBlock: BlockInnerComponent = ({ block }) => {
  const gridColumn = block.grid_column;
  if (!block.children || !Array.isArray(block.children)) {
    return null;
  }

  // width_ratio: ratio of this column to the whole grid (1-99)
  const widthRatio = gridColumn?.width_ratio || 1;

  // Render children by block reference (blockId), allowing for block reference resolution.
  return (
    <div
      className="reark-grid-column"
      style={{
        flex: widthRatio,
        minWidth: 0,
        display: "flex",
        flexDirection: "column",
        gap: "8px",
      }}
    >
      {block.children.map((childId) => (
        // BlockComponent resolves and renders the block by its ID (block reference)
        <BlockComponent key={childId} blockId={childId} />
      ))}
    </div>
  );
};
