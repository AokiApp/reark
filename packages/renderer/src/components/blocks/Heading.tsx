import { BlockInnerComponent } from "../../types";
import { Text } from "./Text";

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

const HeadingBase: BlockInnerComponent<{
  level: HeadingLevel;
}> = ({ block, level }) => {
  const headingKey = `heading${level}` as const;
  const headingData = block[headingKey];

  if (!headingData?.elements) {
    return null;
  }

  return (
    <div
      className={`reark-heading reark-heading--${level}`}
      id={block.block_id}
    >
      <Text elements={headingData.elements} style={headingData.style} />
    </div>
  );
};

export const Heading1: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={1} />
);
export const Heading2: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={2} />
);
export const Heading3: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={3} />
);
export const Heading4: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={4} />
);
export const Heading5: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={5} />
);
export const Heading6: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={6} />
);
export const Heading7: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={7} />
);
export const Heading8: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={8} />
);
export const Heading9: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={9} />
);
