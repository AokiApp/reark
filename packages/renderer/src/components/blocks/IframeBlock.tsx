import React from "react";
import { BlockInnerComponent } from "../../types";

function decodeUrl(url: string): string {
  try {
    return decodeURIComponent(url);
  } catch {
    return url;
  }
}

export const IframeBlock: BlockInnerComponent = ({ block }) => {
  const component = block.iframe?.component;
  if (!component) return null;

  const { iframe_type, url } = component;
  const decodedUrl = decodeUrl(url);

  if (iframe_type === 15) {
    // YouTube埋め込み
    // URL例: https://www.youtube.com/watch?v=xxxx
    // 埋め込み用URLに変換
    let embedUrl = "";
    const match = decodedUrl.match(
      /(?:youtube\.com\/watch\?v=|youtu\.be\/)([A-Za-z0-9_-]+)/,
    );
    if (match && match[1]) {
      embedUrl = `https://www.youtube.com/embed/${match[1]}`;
    } else {
      embedUrl = decodedUrl;
    }
    return (
      <div className="reark-iframe">
        <iframe
          src={embedUrl}
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
          title="YouTube Video"
        />
      </div>
    );
  } else {
    // 汎用iframe
    return (
      <div className="reark-iframe">
        <iframe
          src={decodedUrl}
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
          title="Embedded Content"
        />
      </div>
    );
  }
};
