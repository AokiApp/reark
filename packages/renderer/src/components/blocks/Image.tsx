import { BlockInnerComponent } from "../../types";
import { useContext } from "react";
import { LarkApiContext } from "../../contexts/LarkApiContext";

export const Image: BlockInnerComponent = ({ block }) => {
  const { files } = useContext(LarkApiContext);

  if (!block.image) {
    return null;
  }

  const { width, height, token } = block.image;
  if (!width || !height || !token) {
    return null;
  }

  const align = block.image.align || 1; // Default to left alignment
  const justifyContent =
    ["", "flex-start", "center", "flex-end"][align] || "flex-start";

  const url = files?.[token];

  return (
    <div className="reark-image-container" style={{ justifyContent }}>
      {!url ? (
        <div className="reark-image-placeholder">Image not available</div>
      ) : (
        <img
          src={url}
          alt=""
          className="reark-image"
          width={width}
          height={height}
        />
      )}
    </div>
  );
};
