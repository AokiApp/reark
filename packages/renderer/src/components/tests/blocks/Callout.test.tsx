import { render, screen } from "@testing-library/react";
import { Callout } from "../../blocks/Callout";
import { BlockStoreProvider } from "../../../contexts/BlockStoreContext";

// .private.local/example-blocks.json よりCalloutブロックの例
const calloutBlock = {
  block_id: "PrkUdrRERoRwQfxoLZhjS6qfpFf",
  block_type: 19,
  callout: {
    background_color: 4,
    border_color: 4,
    emoji_id: "innocent",
    text_color: 0
  },
  children: [
    "DRwodrwDoopaGuxAiTUjx2shp3f",
    "G4OOdwJdnoaarOxBhYsjm19LpHb"
  ],
  comment_ids: [
    "7500745586526650402"
  ],
  parent_id: "ZIjadstYfoQVMjxXAwRjM0rVpVg"
};

const heading3Block = {
  block_id: "DRwodrwDoopaGuxAiTUjx2shp3f",
  block_type: 5,
  comment_ids: [
    "7500751511807328290",
    "7500745624808063009"
  ],
  heading3: {
    elements: [
      {
        text_run: {
          content: "コー",
          text_element_style: {
            bold: false,
            inline_code: false,
            italic: false,
            strikethrough: false,
            underline: false
          }
        }
      },
      {
        text_run: {
          content: "ル",
          text_element_style: {
            bold: false,
            comment_ids: ["7500745624808063009"],
            inline_code: false,
            italic: false,
            strikethrough: false,
            underline: false
          }
        }
      },
      {
        text_run: {
          content: "ア",
          text_element_style: {
            bold: false,
            comment_ids: ["7500751511807328290"],
            inline_code: false,
            italic: false,
            strikethrough: false,
            underline: false
          }
        }
      },
      {
        text_run: {
          content: "ウト",
          text_element_style: {
            bold: false,
            inline_code: false,
            italic: false,
            strikethrough: false,
            underline: false
          }
        }
      }
    ],
    style: {
      align: 1 as 1,
      folded: false
    }
  }
};

const bulletBlock = {
  block_id: "G4OOdwJdnoaarOxBhYsjm19LpHb",
  block_type: 12,
  bullet: {
    elements: [
      {
        text_run: {
          content: "ネスト",
          text_element_style: {
            bold: false,
            inline_code: false,
            italic: false,
            strikethrough: false,
            underline: false
          }
        }
      }
    ],
    style: {
      align: 1 as 1,
      folded: false
    }
  }
};

describe("Callout block", () => {
  it("renders callout with emoji and children", () => {
    const { container } = render(
      <BlockStoreProvider items={[calloutBlock, heading3Block, bulletBlock]}>
        <Callout block={calloutBlock} />
      </BlockStoreProvider>
    );
    // Callout emoji
    expect(screen.getByText("😇")).toBeInTheDocument();
    // heading3: "コー", "ル", "ア", "ウト"（まとめて"コールアウト"で部分一致）
    expect(screen.getByText(/コー/)).toBeInTheDocument();
    expect(screen.getByText(/ル/)).toBeInTheDocument();
    expect(screen.getByText(/ア/)).toBeInTheDocument();
    expect(screen.getByText(/ウト/)).toBeInTheDocument();
    // bullet: "ネスト"
    expect(screen.getByText(/ネスト/)).toBeInTheDocument();
    // スナップショットテスト
    expect(container).toMatchSnapshot();
  });
});