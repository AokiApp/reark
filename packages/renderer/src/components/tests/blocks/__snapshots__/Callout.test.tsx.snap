// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Callout block renders callout with emoji and children 1`] = `
<div>
  <div
    class="reark-callout reark-callout--custom"
    style="background-color: rgb(217, 245, 214); border: 1px solid #8ee085; border-radius: 8px;"
  >
    <div
      class="reark-callout__emoji"
    >
      üòá
    </div>
    <div
      class="reark-callout__content"
    >
      <div
        class="reark-comment reark-comment--leaf"
        data-comment-ids="7500751511807328290,7500745624808063009"
      >
        <div
          class="reark-block"
        >
          <div
            class="reark-heading reark-heading--3"
            style="text-align: left;"
          >
            <span
              class=""
            >
              „Ç≥„Éº
            </span>
            <span
              class=""
            >
              „É´
            </span>
            <span
              class=""
            >
              „Ç¢
            </span>
            <span
              class=""
            >
              „Ç¶„Éà
            </span>
          </div>
        </div>
      </div>
      <div
        class="reark-block"
      >
        <div
          class="reark-unordered-list"
        >
          <div
            class="reark-text-block"
            style="text-align: left;"
          >
            <span
              class="reark-text"
            >
              „Éç„Çπ„Éà
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
